File: modules/messaging/src/test/clojure/org/fnbox/messaging/processors/test/QueueConfigurationProcessor.clj
;; Copyright 2008-2011 Red Hat, Inc, and individual contributors.
;; 
;; This is free software; you can redistribute it and/or modify it
;; under the terms of the GNU Lesser General Public License as
;; published by the Free Software Foundation; either version 2.1 of
;; the License, or (at your option) any later version.
;; 
;; This software is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; Lesser General Public License for more details.
;; 
;; You should have received a copy of the GNU Lesser General Public
;; License along with this software; if not, write to the Free
;; Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
;; 02110-1301 USA, or see the FSF site: http://www.fsf.org.

(ns org.fnbox.messaging.processors.test.QueueConfigurationProcessor
  (:use clojure.test)
  (:use fnbox.test.helpers)
  (:use fnbox.test.as.helpers)
  (:import [org.fnbox.core.processors AppCljParsingProcessor])
  (:import [org.fnbox.core ClojureMetaData])
  (:import [org.fnbox.messaging.processors QueueConfigurationProcessor])
  (:import [org.projectodd.polyglot.messaging.destinations QueueMetaData])
  (:require [clojure.java.io :as io]))

(use-fixtures :each
              (harness-with [(AppCljParsingProcessor.)
                             (QueueConfigurationProcessor.)]))

(deftest it-should-not-attach-metadata-if-no-queue-specified
  (let [unit (.deployResourceAs *harness* (io/resource "destinationless-descriptor.clj") "app.clj" )]
    (is (= 0 (.size (.getAttachmentList unit QueueMetaData/ATTACHMENTS_KEY))))))

(deftest it-should-attach-metadata-for-each-queue-specified
  (let [unit (.deployResourceAs *harness* (io/resource "queue-descriptor.clj") "app.clj" )]
    (is (= 3 (.size (.getAttachmentList unit QueueMetaData/ATTACHMENTS_KEY))))))

(defn find-metadata [list name]
  (first (filter #(= (.getName %) name) list)))

(deftest it-should-set-the-name-on-the-metadata
  (let [unit (.deployResourceAs *harness* (io/resource "queue-descriptor.clj") "app.clj" )
        metadatas (.getAttachmentList unit QueueMetaData/ATTACHMENTS_KEY)]
    (are-not [q] (nil? (find-metadata metadatas q))
             "/queue/one"
             "/queue/two"
             "/queue/three")))

(deftest it-should-set-durable-if-requested
  "Pending normalization of descriptor option keys"
  (let [unit (.deployResourceAs *harness* (io/resource "queue-descriptor.clj") "app.clj" )
        metadatas (.getAttachmentList unit QueueMetaData/ATTACHMENTS_KEY)]
    (are [q durable] (= durable (.isDurable (find-metadata metadatas q)))
             "/queue/one" true
             "/queue/two" false)))

(deftest it-should-be-durable-if-not-specified
  (let [unit (.deployResourceAs *harness* (io/resource "queue-descriptor.clj") "app.clj" )
        metadatas (.getAttachmentList unit QueueMetaData/ATTACHMENTS_KEY)]
    (is (= true (.isDurable (find-metadata metadatas "/queue/three"))))))

File: modules/messaging/src/main/java/org/immutant/messaging/processors/QueueConfigurationProcessor.java
/*
 * Copyright 2008-2011 Red Hat, Inc, and individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.immutant.messaging.processors;

import java.util.Map;

import org.immutant.core.ClojureMetaData;
import org.jboss.as.server.deployment.DeploymentPhaseContext;
import org.jboss.as.server.deployment.DeploymentUnit;
import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
import org.jboss.as.server.deployment.DeploymentUnitProcessor;
import org.jboss.logging.Logger;
import org.projectodd.polyglot.messaging.destinations.QueueMetaData;

public class QueueConfigurationProcessor implements DeploymentUnitProcessor {
    
    public QueueConfigurationProcessor() {
    }

    @Override
    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
        DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();

        ClojureMetaData appMetaData = deploymentUnit.getAttachment( ClojureMetaData.ATTACHMENT_KEY );
                
        if (appMetaData == null) {
            return;
        }
        
        Map<String, ?> queues = appMetaData.getHash( "queues" );

        if (queues == null) {
            return;
        }
        
        for(String queue : queues.keySet()) {
            QueueMetaData metaData = new QueueMetaData( queue );
            Map <String, ?> options = (Map<String, Object>)queues.get( queue );
            if (options.containsKey( "durable" )) {
                metaData.setDurable( (Boolean)options.get( "durable" ) );
            }
            deploymentUnit.addToAttachmentList( QueueMetaData.ATTACHMENTS_KEY, metaData );
        }
    }

    @Override
    public void undeploy(DeploymentUnit context) {
       
    }
        
    static final Logger log = Logger.getLogger( "org.immutant.messaging" );
}
